src/test/scala/org/apache/spark/sql/pulsar/PulsarAdmissionControlHelper.scala

```diff
@@ -0,0 +1,88 @@
+package org.apache.spark.sql.pulsar
+
+import java.{util => ju}
+import java.io.Closeable
+import java.util.concurrent.TimeUnit
+import java.util.regex.Pattern
+
+import scala.collection.mutable
+import scala.language.postfixOps
+import scala.util.control.NonFatal
+
+import org.apache.pulsar.client.admin.PulsarAdmin
+import org.apache.pulsar.client.api.{MessageId, PulsarClient}
+import org.apache.pulsar.client.impl.{MessageIdImpl, PulsarClientImpl}
+import org.apache.pulsar.client.impl.schema.BytesSchema
+import org.apache.pulsar.client.internal.DefaultImplementation
+import org.apache.pulsar.common.api.proto.CommandGetTopicsOfNamespace
+import org.apache.pulsar.common.naming.TopicName
+import org.apache.pulsar.common.schema.SchemaInfo
+import org.apache.pulsar.shade.com.google.common.util.concurrent.Uninterruptibles
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.streaming
+import org.apache.spark.sql.connector.read.streaming.{ReadAllAvailable, ReadLimit}
+import org.apache.spark.sql.pulsar.PulsarOptions._
+import org.apache.spark.sql.pulsar.PulsarSourceUtils.{getEntryId, getLedgerId}
+import org.apache.spark.sql.pulsar.SpecificPulsarOffset.getTopicOffsets
+import org.apache.spark.sql.types.StructType
+
+
+class PulsarAdmissionControlHelper(adminUrl: String) {
```

Comment:
You should move the helper to the implementation instead of duplicating the code here.